src/main/java/StringManipulation.java

```diff
@@ -1,33 +1,115 @@
 
     @Override
     public String removeNthCharacter(int n, boolean maintainSpacing) {
-        return null;
+        // if n < 0, throw an error
+        if (n <= 0) {
+            throw new IllegalArgumentException("'n' must be greater than zero.");
+        }
+        // if n > length, throw an error
+        if (n > string.length()) {
+            throw new IndexOutOfBoundsException("'n' is greater than the string length.");
+        }
+        // catching null errors
+        if (string == null) {
+            throw new NullPointerException("String is null!");
+        }
+        // string builder to manipulate
+        StringBuilder result = new StringBuilder();
+        // setting count to 1
+        int count = 1;
+        // looping through the string
+        for (int i = 0; i < string.length(); i++) {
+            if (count % n != 0) {
+                result.append(string.charAt(i));// append char to the result string
+            } else if (maintainSpacing) {
+                result.append(' ');// append a space to maintain spacing
+            }
+            count++;// increment count
+        }
+
+        return result.toString();// return modified string
     }
 
     @Override
     public String[] getSubStrings(int startWord, int endWord) {
-        return null;
+        // catching null string
+        if (string == null) {
+            throw new NullPointerException("String cannot be null!");
+        }
+        String[] words = string.trim().split("\\s+");// splitting the string into words
+        // throwing an error if out of bounds
+        if (startWord <= 0 || endWord <= 0 || startWord > endWord || endWord > words.length) {
+            throw new IllegalArgumentException("Invalid values for startWord and endWord");
+        }
+        // catching end word larger than length of string
+        if (endWord > words.length) {
+            throw new IndexOutOfBoundsException("Not enough words in the string");
+        }
+
+        int length = endWord - startWord + 1;
+        String[] subStrings = new String[length];
+
+        for (int i = startWord - 1; i < endWord; i++) {
+            subStrings[i - startWord + 1] = words[i]; // copying words into array
+        }
+
+        return subStrings;// returning the selected substrings
     }
 
     @Override
     public String restoreString(int[] indices) {
-        return null;
-    }
+        if (string == null || indices.length != string.length()) {
+            throw new IllegalArgumentException("Invalid length of indices array");
+        }
 
+        char[] characters = string.toCharArray(); // Convert the string to a character array
+        char[] shuffled = new char[characters.length]; // Create a new character array for the shuffled string
+        for (int i = 0; i < characters.length; i++) { // Iterate over the indices array and restore the shuffled string
+            if (indices[i] < 0 || indices[i] >= characters.length) { // Check if the index is out of bounds
```

Comment:
no need to check for indices < 0, because trying to access that index will already throw an IndexOutOfBoundsException